### ISSUE TITLE
First field of a struct not recognized appropriately

### ISSUE THREAD

**Issue by ubitux** [NationalSecurityAgency/ghidra#643]
In the following source, the first field of the global struct is not recognized appropriately:

```c
/* first-field.c */
#include <stdio.h>

struct {
    char s0[32];
    char s1[32];
} foo;

int main(int ac, char **av)
{
    if (ac != 3)
        return -1;
    snprintf(foo.s0, sizeof(foo.s0), av[1]);
    snprintf(foo.s1, sizeof(foo.s1), av[2]);
    return 0;
}
```

`gcc -O2 -Wall first-field.c -o first-field`

In Ghidra, it can be reconstructed to:

```c
int main(int ac,char **av)

{
    if (ac == 3) {
        snprintf((char *)&foo,32,av[1]);
        snprintf(foo.s1,32,av[2]);
        return 0;
    }
    return -1;
}
```

The first `snprintf` doesn't seem to recognize that the first field is a string, it should be `snprintf(foo.s0,32,av[1]);`

![2019-05-31-102210_1088x608_scrot](https://user-images.githubusercontent.com/34467/58692466-ef022280-838e-11e9-83c3-6cf9c5baf8db.png)

Env: `Ghidra 9.0.4`, `Linux x86-64`, `GCC 8.3.0`

**Comment by 0x6d696368** [comment]
This is ambiguous. During compilation you lose information that can not be reconstructed during decompilation. E.g. what if your original code actually was:
```C
snprintf((char *)&foo,32,av[1]);
```
Ghidra wouldn't know!

Also what if you have:
```
struct {
union {
    char s0[32];
   char ambiguous[32];
}
    char s1[32];
} foo;
```
Which would it be? `foo.s0` or `foo.ambiguous`?

This is all because you lose information during compilation that can not be recovered during de-compilation.

As an enhancement (in case using the first member of a structure results in less casts) this could create cleaner code and I fully support it. But this is not a bug per se.

**Comment by ubitux** [comment]
> As an enhancement (in case using the first member of a structure results in less casts) this could create cleaner code and I fully support it.

It's more than "less casts" here, it's 0 cast.

> But this is not a bug per se.

Sure, the code is not incorrect in itself, but I consider this the bug because Ghidra is not honoring the non-ambiguity: first field type is known, not a union, and non-ambiguous, therefore re-interpreting the struct instead of pointing to the first field is incorrect (as in: it's not using all the information it's given / honoring user input).

**Comment by nik0sc** [comment]
A raw pointer to the base of `foo` (which is ultimately what Ghidra sees) could just as easily refer to either the entire struct with `sizeof=64` or just s0 with `sizeof=32`. With the second argument to `snprintf` a human could easily infer that the real target is `foo.s0` -- I suspect this is too much to ask of a decompiler. Maybe the best way is to let users resolve the exact type in cases of ambiguity. This would also fix the union problem.

**Comment by caheckman** [comment]
This is addressed by 31b30adf2dfdd2688cc0e4e7cd01d125fc27c45b
